src/cpp/core/FilePath.cpp

```diff
@@ -692,13 +692,36 @@ Error FilePath::move(const FilePath& targetPath) const
    }
    catch(const boost::filesystem::filesystem_error& e)
    {
+      if (type == MoveCrossDevice &&
+          e.code() == boost::system::errc::cross_device_link)
+      {
+         // this error implies that we're trying to move a file from one 
+         // device to another; in this case, fall back to copy/delete
+         return moveIndirect(targetPath);
+      }
       Error error(e.code(), ERROR_LOCATION) ;
       addErrorProperties(pImpl_->path, &error) ;
       error.addProperty("target-path", targetPath.absolutePath()) ;
       return error ;
    }
 }
 
+Error FilePath::moveIndirect(const FilePath& targetPath) const 
+{
+   // copy the file or directory to the new location
+   Error error = isDirectory() ? 
+      copyDirectoryRecursive(targetPath) : copy(targetPath);
```

Comment:
That's a good question. The annoying shell behavior does in fact appear to be what boost's `rename` does, so we should indeed simulate it here for accuracy. Taken care of here: 

https://github.com/rstudio/rstudio/pull/898/commits/144aed5c47f08687caf3403051e8d1eb2c4f0e28